lib/debug/handlers.go

```diff
@@ -0,0 +1,114 @@
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/http/pprof"
+	"strings"
+
+	"github.com/gravitational/teleport/lib/service/servicecfg"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+// NewServeMux returns a http mux that handles all the debug service endpoints.
+func NewServeMux(ctx context.Context, logger *slog.Logger, config *servicecfg.Config) *http.ServeMux {
+	mux := http.NewServeMux()
+	mux.HandleFunc(PProfEndpointsPrefix, pprofMiddleware(ctx, logger, pprof.Index))
+	mux.HandleFunc(PProfEndpointsPrefix+"cmdline", pprofMiddleware(ctx, logger, pprof.Cmdline))
+	mux.HandleFunc(PProfEndpointsPrefix+"profile", pprofMiddleware(ctx, logger, pprof.Profile))
+	mux.HandleFunc(PProfEndpointsPrefix+"symbol", pprofMiddleware(ctx, logger, pprof.Symbol))
+	mux.HandleFunc(PProfEndpointsPrefix+"trace", pprofMiddleware(ctx, logger, pprof.Trace))
+	mux.Handle(LogLevelEndpoint, newLogLevelHandler(ctx, logger, config))
+	return mux
+}
+
+// logLevelHandler http handler of the set/get log level endpoints.
+type logLevelHandler struct {
+	ctx           context.Context
+	logger        *slog.Logger
+	processConfig *servicecfg.Config
+}
+
+func newLogLevelHandler(ctx context.Context, logger *slog.Logger, config *servicecfg.Config) *logLevelHandler {
+	return &logLevelHandler{
+		ctx:           ctx,
+		logger:        logger,
+		processConfig: config,
+	}
+}
+
+func (h *logLevelHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == GetLogLevelMethod {
+		w.Write([]byte(logutils.MarshalText(h.processConfig.LoggerLevel.Level())))
+		return
+	}
+
+	if r.Method != SetLogLevelMethod {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	rawLevel, err := io.ReadAll(io.LimitReader(r.Body, 1024))
+	defer r.Body.Close()
+	if err != nil {
+		w.WriteHeader(http.StatusUnprocessableEntity)
+		w.Write([]byte("Unable to read request body."))
+		return
+	}
+
+	level, err := logutils.UnmarshalText(rawLevel)
+	if err != nil {
+		h.logger.WarnContext(h.ctx, "Failed to parse log level", "error", err)
+		w.WriteHeader(http.StatusUnprocessableEntity)
+		w.Write([]byte("Invalid log level."))
+		return
+	}
+
+	currLevel := h.processConfig.LoggerLevel.Level()
+	message := fmt.Sprintf("Log level already set to %q.", level)
+	if level != currLevel {
+		message = fmt.Sprintf("Changed log level from %q to %q.", currLevel, level)
+		h.processConfig.SetLogLevel(level)
+		h.logger.InfoContext(h.ctx, "Changed log level.", "old", currLevel, "new", string(rawLevel))
+	}
+
+	w.Write([]byte(message))
+}
+
+// pprofMiddleware logs pprof HTTP requests.
+func pprofMiddleware(ctx context.Context, logger *slog.Logger, next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		seconds := r.URL.Query().Get("seconds")
+		if seconds == "" {
+			seconds = "default"
+		}
+
+		logger.InfoContext(
+			ctx,
+			"Collecting pprof profile.",
+			"profile", strings.TrimSuffix(r.URL.Path, PProfEndpointsPrefix),
```

Comment:
That wouldn't cover the `pprof.Index` handler, which accepts the other profiles. I prefer to have a single middleware definition here. (obs: using the new servemux would improve it)

Reply:
> Is it even possible to hit the index profile with the teleport debug command?

Yes. We use the exact definition as [`net/pprof`](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/net/http/pprof/pprof.go;l=93):
- 4 profiles defined by `net/pprof` with their handlers.
- 6 profiles defined by [`runtime/pprof`](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/runtime/pprof/pprof.go;l=227-232) handled by the `Index` handler.

> Doesn't it use our predefined default set of profiles if none are provided?

We copied those definitions because they are not public.

Reply:
I've updated the code to use the go1.22 mux. Let me know if you think we can improve the usage.